build/rewrite.go

```diff
@@ -911,6 +938,245 @@ func fixMultilinePlus(f *File, _ *Rewriter) {
+		// when the file is automatically generated or has automatic transformations
+		// applied to it). The rewrite checks for the comment anywhere in the file
+		// because it's hard to determine which statement is out of order.
+		return
+	}
+
+	// Find the misplaced load statements
+	misplacedLoads := make(map[int]*LoadStmt)
+	firstStmtIndex := -1 // index of the first seen non-load statement
+	for i := 0; i < len(f.Stmt); i++ {
+		stmt := f.Stmt[i]
+		_, isString := stmt.(*StringExpr) // typically a docstring
+		_, isComment := stmt.(*CommentBlock)
+		isWorkspace := isFunctionCall(stmt, "workspace")
+		if isString || isComment || isWorkspace || stmt == nil {
+			continue
+		}
+		load, ok := stmt.(*LoadStmt)
+		if !ok {
+			if firstStmtIndex == -1 {
+				firstStmtIndex = i
+			}
+			continue
+		}
+		if firstStmtIndex == -1 {
+			continue
+		}
+		misplacedLoads[i] = load
+	}
+
+	// Calculate a fix
+	if firstStmtIndex == -1 {
+		firstStmtIndex = 0
+	}
+	offset := len(misplacedLoads)
+	if offset == 0 {
+		return
+	}
+	stmtCopy := append([]Expr{}, f.Stmt...)
+	for i := range f.Stmt {
+		if i < firstStmtIndex {
+			// Docstring or a comment in the beginning, skip
+			continue
+		} else if _, ok := misplacedLoads[i]; ok {
+			// A misplaced load statement, should be moved up to the `firstStmtIndex` position
+			f.Stmt[firstStmtIndex] = stmtCopy[i]
+			firstStmtIndex++
+			offset--
+			if offset == 0 {
+				// No more statements should be moved
+				break
+			}
+		} else {
+			// An actual statement (not a docstring or a comment in the beginning), should be moved
+			// `offset` positions down.
+			f.Stmt[i+offset] = stmtCopy[i]
+		}
+	}
+}
+
+// compressSameOriginLoads merges loads from the same source into one load statement
+func compressSameOriginLoads(f *File, _ *Rewriter) {
+	// Load statements grouped by their source files
+	loads := make(map[string]*LoadStmt)
+
+	for stmtIndex := 0; stmtIndex < len(f.Stmt); stmtIndex++ {
+		load, ok := f.Stmt[stmtIndex].(*LoadStmt)
+		if !ok {
+			continue
+		}
+
+		previousLoad, ok := loads[load.Module.Value]
+		if !ok {
+			loads[load.Module.Value] = load
+			continue
+		}
+		if hasComment(previousLoad, "disable=same-origin-load") ||
+			hasComment(load, "disable=same-origin-load") {
+			continue
+		}
+
+		// Move loaded symbols to the existing load statement
+		previousLoad.From = append(previousLoad.From, load.From...)
+		previousLoad.To = append(previousLoad.To, load.To...)
+		// Remove the current statement
+		f.Stmt[stmtIndex] = nil
+	}
+}
+
+// comparePaths compares two strings as if they were paths (the path delimiter,
+// '/', should be treated as smallest symbol).
+func comparePaths(path1, path2 string) bool {
+	if path1 == path2 {
+		return false
+	}
+
+	chunks1 := strings.Split(path1, "/")
+	chunks2 := strings.Split(path2, "/")
+
+	for i, chunk1 := range chunks1 {
+		if i >= len(chunks2) {
+			return false
+		}
+		chunk2 := chunks2[i]
+		// Compare case-insensitively
```

Comment:
are bzl paths not case sensitive generally?  is `build` as good as `BUILD`

Reply:
Depends on the file system. On Linux the paths are usually case sensitive, `build` and `BUILD` can be two separate files in ext3.